rust/src/traits/samplers/cks20/sample_discrete_laplace.tex

```diff
@@ -1,112 +1,109 @@
-
-\begin{lemma}\label{ok-out}
-    If the outcome of \texttt{sample\_discrete\_laplace} is \texttt{Ok(out)}, 
-    then \texttt{out} is distributed as $\mathcal{L}_\mathbb{Z}(0, scale)$.
-\end{lemma}
-
-\begin{proof}
-    In the \ref{err-e} proof, it was established that the preconditions on \texttt{sample\_geometric\_exp\_fast} are met.
-    therefore \texttt{magnitude} on line \ref{line:magnitude} is distributed as $ Geometric(1 - e^{-1/scale})$.
-    Similarly, by the definition of \texttt{sample\_standard\_bernoulli}, \texttt{sign} is distributed according to $Bernoulli(p=1/2)$.
-    The branching logic from line \ref{line:branching} on satisfies the procedures described in \ref{P_Lx_BY_ne_T0}.
-    Therefore, by \ref{P_Lx_BY_ne_T0}, \texttt{out} is distributed as $\mathcal{L}_\mathbb{Z}(0, scale)$.
-\end{proof}
-
-\begin{proof}
-    \ref{postcondition} holds by \ref{err-e} and \ref{ok-out}.
-\end{proof}
-
-
-\bibliographystyle{alpha}
-\bibliography{mod}
-
+\documentclass{article} 
+\input{../../../lib.sty} 
+ 
+\title{\texttt{fn sample\_discrete\_laplace}} 
+\author{Michael Shoemate} 
+ 
+\begin{document} 
+\maketitle 
+ 
+Proves soundness of \texttt{fn sample\_discrete\_laplace} in \asOfCommit{mod.rs}{0be3ab3e6}. 
+This proof is an adaptation of \href{https://arxiv.org/pdf/2004.00010.pdf#subsection.5.2}{subsection 5.2} of \cite{CKS20}. 
+ 
+\section{Hoare Triple} 
+\subsection*{Precondition} 
+\textbf{Compiler-verified} 
+Argument \texttt{scale} is of type \texttt{RBig}, a bignum rational 
+ 
+\textbf{User-verified} 
+$\texttt{scale} \geq 0$ 
+ 
+\subsection*{Pseudocode}         
+\lstinputlisting[language=Python,firstline=2,escapechar=|]{./pseudocode/sample_discrete_laplace.py} 
+ 
+\subsection*{Postcondition} 
+\label{postcondition} 
+For any setting of the input parameter \texttt{scale} such that the given preconditions hold, \\ 
+\texttt{sample\_discrete\_laplace} either returns \texttt{Err(e)} due to a lack of system entropy, 
+or \texttt{Ok(out)}, where \texttt{out} is distributed as $\mathcal{L}_\mathbb{Z}(0, scale)$. 
+ 
+\section{Proof} 
+\begin{definition} \cite{BV17} 
+    (Discrete Laplace). Let $\mu, \sigma \in \mathbb{R}$ with $\sigma > 0$.  
+    The discrete laplace distribution with location $\mu$ and scale $s$ is denoted $\mathcal{L}_\mathbb{Z}(\mu, s)$.  
+    It is a probability distribution supported on the integers and defined by 
+    \begin{equation*} 
+        \forall x \in \mathbb{Z} \quad  P[X = x] = \frac{e^{1/s} - 1}{e^{1/s} + 1} e^{-|x|/s} \quad \text{where } X \sim \mathcal{L}_\mathbb{Z}(\mu, s) 
+    \end{equation*} 
+\end{definition} 
+ 
+Assume the preconditions are met. 
+ 
+\begin{lemma} 
+    \label{err-e} 
+    \texttt{sample\_discrete\_laplace} only returns \texttt{Err(e)} when there is a lack of system entropy. 
+\end{lemma} 
+ 
+\begin{proof} 
+    By the non-negativity precondition on \texttt{scale}, 
+    the precondition on \rustdoc{traits/samplers/cks20/fn}{sample\_geometric\_exp\_fast} is met. 
```

Comment:
I've split the preconditions for that proof into compiler-verified (scale is rational) and user-verified (scale is positive). `scale` being non-negative is user-verified, as we've done here. For brevity, I think we should skip compiler-verified preconditions. This will be made clearer once that proof is updated.